executor/containerdexecutor/executor_unix.go

```diff
@@ -0,0 +1,170 @@
+//go:build !windows
+// +build !windows
+
+package containerdexecutor
+
+import (
+	"context"
+	"os"
+	"runtime"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/mount"
+	containerdoci "github.com/containerd/containerd/oci"
+	"github.com/containerd/continuity/fs"
+	"github.com/docker/docker/pkg/idtools"
+	"github.com/moby/buildkit/executor"
+	"github.com/moby/buildkit/executor/oci"
+	"github.com/moby/buildkit/snapshot"
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/moby/buildkit/util/network"
+	rootlessspecconv "github.com/moby/buildkit/util/rootless/specconv"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+func getUserSpec(user, rootfsPath string) (specs.User, error) {
+	var err error
+	var uid, gid uint32
+	var sgids []uint32
+	if rootfsPath != "" {
+		uid, gid, sgids, err = oci.GetUser(rootfsPath, user)
+	} else {
+		uid, gid, err = oci.ParseUIDGID(user)
+	}
+	if err != nil {
+		return specs.User{}, errors.WithStack(err)
+	}
+	return specs.User{
+		UID:            uid,
+		GID:            gid,
+		AdditionalGids: sgids,
+	}, nil
+}
+
+func (w *containerdExecutor) prepareExecutionEnv(ctx context.Context, rootMount executor.Mount, mounts []executor.Mount, meta executor.Meta, details *jobDetails) (func(), string, string, error) {
+	var releasers []func()
+	releaseAll := func() {
+		for i := len(releasers) - 1; i >= 0; i-- {
+			releasers[i]()
+		}
+	}
+
+	resolvConf, err := oci.GetResolvConf(ctx, w.root, nil, w.dnsConfig)
+	if err != nil {
+		return releaseAll, "", "", err
+	}
+
+	hostsFile, clean, err := oci.GetHostsFile(ctx, w.root, meta.ExtraHosts, nil, meta.Hostname)
+	if err != nil {
+		return releaseAll, "", "", err
+	}
+	if clean != nil {
+		releasers = append(releasers, clean)
+	}
+	mountable, err := rootMount.Src.Mount(ctx, false)
+	if err != nil {
+		return releaseAll, "", "", err
+	}
+
+	rootMounts, release, err := mountable.Mount()
+	if err != nil {
+		return releaseAll, "", "", err
+	}
+	details.rootMounts = rootMounts
+
+	if release != nil {
+		releasers = append(releasers, func() {
+			if err := release(); err != nil {
+				bklog.G(ctx).WithError(err).Error("failed to release root mount")
+			}
+		})
+	}
+	lm := snapshot.LocalMounterWithMounts(rootMounts)
+	rootfsPath, err := lm.Mount()
+	if err != nil {
+		return releaseAll, "", "", err
+	}
+	details.rootfsPath = rootfsPath
+	releasers = append(releasers, func() {
+		if err := lm.Unmount(); err != nil {
+			bklog.G(ctx).WithError(err).Error("failed to unmount rootfs")
+		}
+	})
+	releasers = append(releasers, executor.MountStubsCleaner(ctx, details.rootfsPath, mounts, meta.RemoveMountStubsRecursive))
+
+	return releaseAll, resolvConf, hostsFile, nil
+}
+
+func (w *containerdExecutor) getTaskOpts(ctx context.Context, details *jobDetails) (containerd.NewTaskOpts, error) {
```

Comment:
This can just be a method on `containerState` (jobDetails), same for `cwd`. Not sure if makes sense for others.